From c17ae85ef66f04df514ce05fe41e8202f76c3835 Mon Sep 17 00:00:00 2001
From: Kirti Wankhede <kwankhede@nvidia.com>
Date: Thu, 17 Nov 2016 02:16:18 +0530
Subject: [PATCH 57/64] vfio iommu type1: Update arguments of vfio_lock_acct

Git-commit: 3624a2486c8ca10a2c730c704441fdd034a0d9b7
Patch-mainline: v4.10-rc1
References: fate#321947

Added task structure as input argument to vfio_lock_acct() function.

Signed-off-by: Kirti Wankhede <kwankhede@nvidia.com>
Signed-off-by: Neo Jia <cjia@nvidia.com>
Signed-off-by: Alex Williamson <alex.williamson@redhat.com>
Signed-off-by: Mian Yousaf Kaukab <yousaf.kaukab@suse.com>
---
 drivers/vfio/vfio_iommu_type1.c |   24 ++++++++++++++----------
 1 file changed, 14 insertions(+), 10 deletions(-)

--- a/drivers/vfio/vfio_iommu_type1.c
+++ b/drivers/vfio/vfio_iommu_type1.c
@@ -130,32 +130,36 @@ static void vfio_unlink_dma(struct vfio_
 	rb_erase(&old->node, &iommu->dma_list);
 }
 
-static int vfio_lock_acct(long npage, bool *lock_cap)
+static int vfio_lock_acct(struct task_struct *task, long npage, bool *lock_cap)
 {
+	struct mm_struct *mm;
 	int ret = 0;
 
 	if (!npage)
 		return 0;
 
-	if (!current->mm)
+	mm = get_task_mm(task);
+	if (!mm)
 		return -ESRCH; /* process exited */
 
-	down_write(&current->mm->mmap_sem);
+	down_write(&mm->mmap_sem);
 	if (npage > 0) {
 		if (lock_cap ? !*lock_cap : !capable(CAP_IPC_LOCK)) {
 			unsigned long limit;
 
-			limit = rlimit(RLIMIT_MEMLOCK) >> PAGE_SHIFT;
+			limit = task_rlimit(task,
+					RLIMIT_MEMLOCK) >> PAGE_SHIFT;
 
-			if (current->mm->locked_vm + npage > limit)
+			if (mm->locked_vm + npage > limit)
 				ret = -ENOMEM;
 		}
 	}
 
 	if (!ret)
-		current->mm->locked_vm += npage;
+		mm->locked_vm += npage;
 
-	up_write(&current->mm->mmap_sem);
+	up_write(&mm->mmap_sem);
+	mmput(mm);
 
 	return ret;
 }
@@ -287,7 +291,7 @@ static long vfio_pin_pages_remote(unsign
 
 out:
 	if (!rsvd)
-		ret = vfio_lock_acct(i, &lock_cap);
+		ret = vfio_lock_acct(current, i, &lock_cap);
 
 unpin_out:
 	if (ret) {
@@ -312,7 +316,7 @@ static long vfio_unpin_pages_remote(unsi
 		unlocked += put_pfn(pfn++, prot);
 
 	if (do_accounting)
-		vfio_lock_acct(-unlocked, NULL);
+		vfio_lock_acct(current, -unlocked, NULL);
 
 	return unlocked;
 }
@@ -374,7 +378,7 @@ static void vfio_unmap_unpin(struct vfio
 		cond_resched();
 	}
 
-	vfio_lock_acct(-unlocked, NULL);
+	vfio_lock_acct(current, -unlocked, NULL);
 }
 
 static void vfio_remove_dma(struct vfio_iommu *iommu, struct vfio_dma *dma)
